features/storage/kvstore/DirectAccessDevicekey.cpp

```diff
@@ -0,0 +1,189 @@
+                            tdbstore_area_data_t *area_params);
+static int reserved_data_get(BlockDevice *bd, tdbstore_area_data_t *area_params, void *reserved_data_buf,
+                             size_t reserved_data_buf_size, size_t *actual_data_size_ptr);
+static uint32_t calc_crc(uint32_t init_crc, uint32_t data_size, const void *data_buf);
+
+// -------------------------------------------------- API Functions Implementation ----------------------------------------------------
+int direct_access_to_devicekey(BlockDevice *bd, uint32_t tdb_start_offset, uint32_t tdb_end_offset, void *data_buf,
+                               size_t data_buf_size, size_t *actual_data_size_ptr)
+{
+    int status = MBED_ERROR_INVALID_ARGUMENT;
+    uint8_t active_area = 0;
+    tdbstore_area_data_t area_params[TDBSTORE_NUMBER_OF_AREAS];
+    memset(area_params, 0, sizeof(area_params));
+
+    if (NULL == data_buf) {
+        tr_error("Invalid Data Buf Argument");
+        goto exit_point;
+    }
+
+    status = calc_area_params(bd, tdb_start_offset, tdb_end_offset, area_params);
+    if (status != MBED_SUCCESS) {
+        tr_error("Couldn't calulate Area Params - err: %d", status);
+        goto exit_point;
+    }
+
+    /* DeviceKey data can be found either in first or second Flash Area */
+    /* Loop through Areas to find valid DeviceKey data */
+    for (active_area = 0; active_area < TDBSTORE_NUMBER_OF_AREAS; active_area++) {
+        status = reserved_data_get(bd, &area_params[active_area], data_buf, data_buf_size, actual_data_size_ptr);
+        if (status == MBED_SUCCESS) {
+            break;
+        }
+    }
+
+    if (status != MBED_SUCCESS) {
+        status = MBED_ERROR_ITEM_NOT_FOUND;
+        tr_error("Couldn't find valid DeviceKey - err: %d", status);
+    }
+
+exit_point:
+
+    return status;
+}
+
+// -------------------------------------------------- Local Functions Implementation ----------------------------------------------------
+static int calc_area_params(BlockDevice *bd, uint32_t tdb_start_offset, uint32_t tdb_end_offset,
+                            tdbstore_area_data_t *area_params)
+{
+    bd_size_t bd_size = 0;
+    bd_size_t inital_erase_size = bd->get_erase_size(tdb_start_offset);
+    bd_size_t erase_unit_size = inital_erase_size;
+    size_t cur_area_size = 0;
+
+    if ( (tdb_end_offset < (tdb_start_offset + 2 * RESERVED_AREA_SIZE - 1)) || (tdb_end_offset > bd->size()) ) {
+        tr_error();
+        return MBED_ERROR_INVALID_ARGUMENT;
+    }
+
+    // Entire TDBStore can't exceed 32 bits
+    bd_size = (bd_size_t)(tdb_end_offset - tdb_start_offset + 1);
+
+    while (cur_area_size < bd_size / 2) {
+        erase_unit_size = bd->get_erase_size(tdb_start_offset + cur_area_size);
+        cur_area_size += erase_unit_size;
+    }
+
+    area_params[0].address = tdb_start_offset;
+    area_params[0].size = cur_area_size;
+    area_params[1].address = cur_area_size;
```

Comment:
This seems like a bug. Shouldn't it be `tdb_start_offset + cur_area_size`?

Reply:
indeed a bug - thanks!